build.sbt

```diff
@@ -389,22 +391,33 @@ lazy val commonSettings = Seq(
       case _ => "1.8"
     }
   },
-  scalacOptions <<= jvmTarget.map { jvm => Seq(
+  scalacOptions := Seq(
     "-deprecation",
+    "-encoding", "UTF-8",
     "-feature",
-    "-language:implicitConversions",
+    "-language:existentials",
     "-language:higherKinds",
-    s"-target:jvm-$jvm",
+    "-language:implicitConversions",
+    s"-target:jvm-${jvmTarget.value}",
     "-unchecked",
-    "-Xlint"
-  )},
+    "-Xfatal-warnings",
+    "-Xlint",
+    "-Yinline-warnings",
+    "-Yno-adapted-args",
+    "-Ywarn-dead-code",
+    "-Ywarn-numeric-widen",
+    "-Ywarn-value-discard",
+    "-Xfuture"
+  ),
+  /* disabled because SI-7529
```

Comment:
I know this gives us smaller jars.  I don't know what other speedups it brings.  I don't fully understand the three flags, or how they interact.  Would appreciate some feedback here from someone knowledgeable.

With Scala 2.12 looming, might we get away with dropping these if the Java 8 backend is more mature in Scala 2.12?